server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -540,10 +543,9 @@ private static void parseArray(ParseContext context, ObjectMapper parentMapper,
             Tuple<Integer, ObjectMapper> parentMapperTuple = getDynamicParentMapper(context, paths, parentMapper);
             parentMapper = parentMapperTuple.v2();
             ObjectMapper.Dynamic dynamic = dynamicOrDefault(parentMapper, context);
-            if (dynamic == ObjectMapper.Dynamic.STRICT) {
-                throw new StrictDynamicMappingException(parentMapper.fullPath(), arrayFieldName);
-            } else if (dynamic == ObjectMapper.Dynamic.TRUE) {
-                Mapper.Builder builder = context.root().findTemplateBuilder(context, arrayFieldName, XContentFieldType.OBJECT);
+            if (dynamic.canCreateDynamicFields()) {
```

Comment:
Could we have this return the DynamicFieldsBuilder` or something like it? It kind of feels weird to have the way we handle dynamic fields controlled both by this enum and a pluggable behavior layer.

Reply:
the idea is that this new method on the enum removes the need for the `if (dynamic == RUNTIME || dynamic == TRUE)` all over the place. I could see how we would forget for instance RUNTIME in some places while this way it is streamlined. At the same time, we don't want to make other dynamic modes pluggable, and actually it is only sort of pluggable for the  runtime mode, in that the mapping between the runtime fields and their dynamic type needs to be provided by the plugin. Does that make sense? I am not sure what you were proposing though.

Reply:
I'm sort of wondering if the enum can build the `DynamicFieldsBuilder` and the strict one can throw and the `false` one can noop.

Reply:
I see, I briefly also considered this, and then I went for not making the enum bigger. I will have a look if this simplifies things.